### DISCUSSION TITLE
Anybody run into the need to average over Z axis (e.g., `cdutil.averager(tvar, axis="z")`), and what did you use?

### DISCUSSION THREAD

**Discussion by tomvothecoder** [xCDAT/xcdat#591] [2. Questions] [answered]
### Question criteria

- [X] I added a descriptive title here.
- [X] I searched the [xCDAT GitHub Discussions](https://github.com/xCDAT/xcdat/discussions) to find a similar question and didn't find it.
- [X] I searched the [xCDAT documentation](https://xcdat.readthedocs.io/en/stable/) and/or [Xarray documentation](https://docs.xarray.dev/en/stable/).

### Describe your question

I'm running into a case where I need to average over the Z axis. The CDAT code being used is `cdutil.averager(tvar, axis="z")`, which I need to replace.

@xCDAT/core-developers, has anybody else found the need to average over the Z axis? And if so, what package or custom code are you using to do this?

xCDAT's [spatial averager](https://xcdat.readthedocs.io/en/stable/generated/xarray.Dataset.spatial.average.html#xarray.Dataset.spatial.average), which is based on `cdutil.averager()`, only supports rectilinear grids ("X" and "Y"). However, we do mention that the `get_weights()` method can be extended to support other axes:
https://github.com/xCDAT/xcdat/blob/fbf1db68fd6328d5a534e0c1e57e22e6fdd70c76/xcdat/spatial.py#L246-L252

### Are there are any possible answers you came across?

For the short-term, I might be able to use [xarray.DataArray.weighted](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.weighted.html) directly. However, I still need to generate the weights xr.DataArray beforehand.

```python
datarray.weighted(weights).mean("height")
```

**Comment by pochedls** [comment, marked as answer]
@tomvothecoder – I have not done this recently, but I think the same algorithm would be valid for depth / pressure coordinates (weights should generally scale as the difference between the bound values for each cell). I think it would make sense to extend `get_weights` to be able to handle vertical axes.

**Comment by lee1043** [comment]
@tomvothecoder -- I think PMP is now running into the need of average over Z axis as a part of its ENSO metrics migration to xCDAT from CDAT. Without recalling this discussion was already posted, I advised my colleague to try `spatial.average` with `axis=["Z"]` which of course did not work. 

```python
filename = "/bdd/CMIP6/CMIP/IPSL/IPSL-CM6A-LR/historical/r1i1p1f1/Omon/thetao/gn/latest/thetao_Omon_IPSL-CM6A-LR_historical_r1i1p1f1_gn_185001-194912.nc"
ds = xcdat.open_mfdataset(filename, add_bounds=["T", "X", "Y", "Z"])
ds_ave = ds.spatial.average("thetao", axis=["Z"])
```
Which returns the expected error:
```
ValueError: Incorrect `axis` argument value. Supported values include: X, Y.
``` 

I think it might worth to tackle this. Maybe additional item for the Dev Day?

This might also requires generation of weighting if not provided.

**Reply by pochedls** [reply]
I think this is done in this [PR](https://github.com/xCDAT/xcdat/pull/606), but it needs testing.
